src/active-response/firewalls/pf.c

```diff
@@ -0,0 +1,164 @@
+        return OS_INVALID;
+    }
+
+    // OpenBSD and FreeBSD pf
+    if (!strcmp("OpenBSD", uname_buffer.sysname) || !strcmp("FreeBSD", uname_buffer.sysname) || !strcmp("Darwin", uname_buffer.sysname)) {
+
+        // Checking if pfctl is present
+        if (access(PFCTL, F_OK) < 0) {
+            memset(log_msg, '\0', LOGSIZE);
+            snprintf(log_msg, LOGSIZE - 1, "The pfctl file '%s' is not accessible", PFCTL);
+            write_debug_file(argv[0], log_msg);
+            return OS_SUCCESS;
+        }
+
+        char *exec_cmd1[7];
+        char *exec_cmd2[4];
+
+        // Checking if we have pf config file
+        if(access(PFCTL_RULES, F_OK) == 0) {
+            // Checking if ossec table is configured in pf.conf
+            if(checking_if_its_configured(PFCTL_RULES, PFCTL_TABLE) == 0) {
+                if (!strcmp("add", action)) {
+                    char *arg1[7] = {PFCTL, "-t", PFCTL_TABLE, "-T", "add", srcip, NULL};
+                    memcpy(exec_cmd1, arg1, sizeof(exec_cmd1));
+
+                    char *arg2[4] = {PFCTL, "-k", srcip, NULL};
+                    memcpy(exec_cmd2, arg2, sizeof(exec_cmd2));
+                } else {
+                    char *arg1[7] = {PFCTL, "-t", PFCTL_TABLE, "-T", "delete", srcip, NULL};
+                    memcpy(exec_cmd1, arg1, sizeof(exec_cmd1));
+                }
+            } else {
+                memset(log_msg, '\0', LOGSIZE);
+                snprintf(log_msg, LOGSIZE - 1, "Table %s does not exist", PFCTL_TABLE);
+                write_debug_file(argv[0], log_msg);
+                return OS_INVALID;
+            }
+
+        } else {
+            memset(log_msg, '\0', LOGSIZE);
+            snprintf(log_msg, LOGSIZE - 1, "The pf rules file %s does not exist", PFCTL_RULES);
+            write_debug_file(argv[0], log_msg);
+            return OS_SUCCESS;
+        }
+
+        // Executing it
+        if(strcmp(exec_cmd1[0], PFCTL) == 0) {
+            wfd_t *wfd = wpopenv(PFCTL, exec_cmd1, W_BIND_STDOUT);
+            if(!wfd) {
+                memset(log_msg, '\0', LOGSIZE);
+                snprintf(log_msg, LOGSIZE - 1, "Error executing %s : %s", PFCTL, strerror(errno));
+                write_debug_file(argv[0], log_msg);
+                return OS_INVALID;
+            }
+            wpclose(wfd);
+        }
+
+        if(strcmp(exec_cmd2[0], PFCTL) == 0) {
+            wfd_t *wfd = wpopenv(PFCTL, exec_cmd2, W_BIND_STDOUT);
+            if(!wfd) {
+                memset(log_msg, '\0', LOGSIZE);
+                snprintf(log_msg, LOGSIZE - 1, "Error executing %s : %s", PFCTL, strerror(errno));
+                write_debug_file(argv[0], log_msg);
+                return OS_INVALID;
+            }
+            wpclose(wfd);
+        }
+
+    } else {
+        return OS_SUCCESS;
+    }
+
+    write_debug_file(argv[0], "Ended");
+
+    return 0;
+}
+
+int checking_if_its_configured(const char *path, const char *table) {
+    char command[1023];
```

Comment:
Use `COMMANDSIZE` here and in all the cases below. Use `COMMANDSIZE -1` in `snprintf`.

Reply:
Done